app/v2/controllers/stream/StreamingParsers.scala

```diff
@@ -80,4 +86,99 @@ trait StreamingParsers {
       }
   }
 
+  private val START_TOKEN: Source[ByteString, NotUsed]                 = Source.single(ByteString("{"))
+  private val END_TOKEN_SOURCE: Source[ByteString, NotUsed]            = Source.single(ByteString("}"))
+  private val END_TOKEN_WITH_QUOTE_SOURCE: Source[ByteString, NotUsed] = Source.single(ByteString("\"}"))
+
+  def mergeStreamIntoJson(
+    fields: collection.Seq[(String, JsValue)],
+    fieldName: String,
+    stream: Source[ByteString, _]
+  ): EitherT[Future, PresentationError, Source[ByteString, _]] =
+    EitherT {
+      Future
+        .successful(
+          Right( // We need to start the document and add the final field, and prepare for the string data coming in
+            START_TOKEN ++
+              Source
+                .fromIterator(
+                  () => fields.iterator
+                )
+                .map {
+                  // convert fields to bytestrings with their values, need to ensure we keep the field names quoted,
+                  // separate with a colon and end with a comma, as per the Json spec
+                  tuple =>
+                    ByteString(s""""${tuple._1}":${Json.stringify(tuple._2)},""")
+                } ++
+              // start adding the new field
+              Source.single(ByteString(s""""$fieldName":"""".stripMargin)) ++
+              // our XML stream that we escape
+              stream.map(
+                bs => ByteString(bs.utf8String.replace("\\", "\\\\").replace("\"", "\\\""))
+              ) ++
+              // and the stream that ends our Json document
+              END_TOKEN_WITH_QUOTE_SOURCE
+          )
+        )
+        .recover {
+          case NonFatal(e) => Left(PresentationError.internalServiceError(cause = Some(e)))
+        }
+    }
+
+  def jsonToByteStringStream(fields: collection.Seq[(String, JsValue)]): EitherT[Future, PresentationError, Source[ByteString, _]] =
+    EitherT {
+      Future
+        .successful(
+          Right(
+            START_TOKEN ++ Source
+              .fromIterator(
+                () => fields.iterator
+              )
+              .map {
+                // convert fields to bytestrings with their values, need to ensure we keep the field names quoted,
+                // separate with a colon and end with a comma, as per the Json spec
+                tuple =>
+                  ByteString(s""""${tuple._1}":${Json.stringify(tuple._2)}""")
+              }
+              .intersperse(ByteString(",")) ++ END_TOKEN_SOURCE
+          )
+        )
+        .recover {
+          case NonFatal(e) => Left(PresentationError.internalServiceError(cause = Some(e)))
+        }
+    }
+
+  def jsonWrappedXmlToByteStringStream(fields: collection.Seq[(String, JsValue)]): EitherT[Future, PresentationError, Source[ByteString, _]] =
```

Comment:
Can we make the code within .successful(..) a function as seems duplicated in the 3 functions

Reply:
Sure, I'll reduce the duplication